improve like/share buttons

xundeenergie opened the pull request:
change the button only if it could send its request successfully to the server.

fixes:
* disrupting a video on liking or sharing
* timeline jumps around somewhere else, when you like or share a posting/comment, and you can not find back.

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,52 @@ function htmlToText(htmlString) {
+          console.log("announce")
+          if (data.verb == "un" + verb) {
+            $("button[id^=shareMenuOptions-" + ident.toString() + "]" ).removeClass("active");
+            $("button[id^=shareMenuOptions-" + ident.toString() + "]" ).attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
```
These structures can be simplified, please change the other cases.
```suggestion
            $("button[id^=shareMenuOptions-" + ident.toString() + "]" )
                .removeClass("active")
                .attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,52 @@ function htmlToText(htmlString) {
+            $("button[id^=shareMenuOptions-" + ident.toString() + "]" ).removeClass("active");
+            $("button[id^=shareMenuOptions-" + ident.toString() + "]" ).attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
+            $("button[id^=" + verb + "-" + ident.toString() + "]" ).removeClass("active");
+            $("button[id^=" + verb + "-" + ident.toString() + "]" ).attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
```
In this case we want the boolean string because it's used as a parameter value. Please fix this in the other cases.
```suggestion
            $("button[id^=" + verb + "-" + ident.toString() + "]" ).attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', false)").change();
```

xundeenergie replied:
prove me wrong... 

But i use it as boolean. This was my intention.
In your suggestion the quoting is wrong.

MrPetovan replied:
Just look at how it is written in the `onclick` attribute of the current HTML of the action button:
```html
<button type="button" class="btn-link button-likes active" aria-pressed="true" id="like-5705380" title="I like this (toggle)" onclick="doActivityItemAction(5705380, 'like', true);" data-toggle="button"><i class="fa fa-thumbs-up" aria-hidden="true"></i>&nbsp;Like</button>
```
The boolean parameter value is spelled out `true` or `false`. Since you are also setting the `onclick` attribute value to a string in your code, this parameter value needs to be a string.

And my quoting is correct, I believe you may be confused by the nearby single quote for the `verb` parameter.

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,52 @@ function htmlToText(htmlString) {
-		$(".item-" + ident + " .button-event:not(#" + verb + "-" + ident + ")").removeClass("active");
-	}
-	$("#" + verb + "-" + ident).toggleClass("active");
+  console.log(ident, verb, un);
```
Please remove debug output from your PRs.

xundeenergie replied:
ok. is fixed.

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,52 @@ function htmlToText(htmlString) {
+	_verb = un ? 'un' + verb : verb;
+	$('#like-rotator-' + ident.toString()).show();
+	$.post('item/' + ident.toString() + '/activity/' + _verb)
+	.success(
```
Please indent your changes with tabulations.

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,52 @@ function htmlToText(htmlString) {
+            $("button[id^=" + verb + "-" + ident.toString() + "]" ).removeClass("active");
+            $("button[id^=" + verb + "-" + ident.toString() + "]" ).attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
+          } else {
+            $("button[id^=shareMenuOptions-" + ident.toString() + "]" ).addClass("active");
```
In general, for asynchronous actions, you don't want to wait for the response to change the status of the interactive element. If there's no immediate feedback like a page change or a state change (the rotator isn't enough), the user may think the action didn't go through and they may click again.

Here's a better workflow:
1. User interacts with element.
2. Element changes state immediately.
3. Asynchronous request is performed.
4. On response, the display of the element is updated again depending on the success of the request.

This means that sometimes the display will flicker if the action wasn't successful, and it's okay on its own, we just need an additional piece of feedback that the action failed like you did in the `error` callback.

xundeenergie commented:
Im not used to use github pull-requests...
i updated and pushed the changes to my github-repo.
What should i do now, to publish the new pull-request?

MrPetovan commented:
GitHub picks up the new commits as soon as you push them.

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,50 @@ function htmlToText(htmlString) {
+          } else {
+            $("button[id^=" + verb + "-" + ident.toString() + "]" )
+              .addClass("active")
+              .attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + true + ")").change();
```
In general, for asynchronous actions, you don't want to wait for the response to change the status of the interactive element. If there's no immediate feedback like a page change or a state change (the rotator isn't enough), the user may think the action didn't go through and they may click again.

Here's a better workflow:
1. User interacts with element.
2. Element changes state immediately.
3. Asynchronous request is performed.
4. On response, the display of the element is updated again depending on the success of the request.

This means that sometimes the display will flicker if the action wasn't successful, and it's okay on its own, we just need an additional piece of feedback that the action failed like you did in the `error` callback.

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+							if (data.verb == "un" + verb) {
+									$("button[id^=shareMenuOptions-" + ident.toString() + "]" )
+										.removeClass("active")
+										.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
```
```suggestion
										.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', false)")
										.change();
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+										.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
+									$("button[id^=" + verb + "-" + ident.toString() + "]" )
+										.removeClass("active")
+										.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
```
```suggestion
										.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', false)").change();
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+							} else {
+									$("button[id^=shareMenuOptions-" + ident.toString() + "]" )
+											.addClass("active")
+											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + true + ")").change();
```
```suggestion
											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', true)")
											.change();
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + true + ")").change();
+									$("button[id^=" + verb + "-" + ident.toString() + "]" )
+											.addClass("active")
+											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + true + ")").change();
```
```suggestion
											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', true)")
											.change();
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+							if (data.verb == "un" + verb) {
+									$("button[id^=" + verb + "-" + ident.toString() + "]" )
+											.removeClass("active")
+											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + false + ")").change();
```
```suggestion
											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', false)")
											.change();
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+							} else {
+									$("button[id^=" + verb + "-" + ident.toString() + "]" )
+											.addClass("active")
+											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + true + ")").change();
```
```suggestion
											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', true)")
											.change();
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+											.addClass("active")
+											.attr("onclick", "doActivityItemAction(" + ident +", '" + verb + "', " + true + ")").change();
+							}
+	//      		  $("button[id^=" + verb + "-" + ident.toString() + "]" ).button('refresh');
```
Please remove commented out code.
```suggestion
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,49 @@ function htmlToText(htmlString) {
+	$.post('item/' + ident.toString() + '/activity/' + _verb)
+	.success(
+		function(data){
+			if (data.status == "ok") {
```
Please use single quotes when double-quotes aren't warranted.
```suggestion
			if (data.status == 'ok') {
```

xundeenergie commented:
With the last commit i cleaned up and simplified the code, and added some improvements for optical feedback of transaction on the buttons itself (not in a separat rotator-icon...)

Commit:
simplify code, remove debug output

Commit:
improve like/share buttons

change the button only if it could send its request successfully to the server.

fixes:
* disrupting a video on liking or sharing
* timeline jumps around somewhere else, when you like or share a
  posting/comment, and you can not find back.

Commit:
Merge pull request #12875 from annando/simplify-oauth

Use the built in function to create a query string

Commit:
Comment readded

Commit:
improve optical feedback with rotator on buttons

* functional rotator with feedback from ajax-request for activity
    on like, dislike, share and shareMenu button in wide-screen and media-screen

Commit:
show rotator on button

Commit:
Use the built in function to create a query string

Commit:
make indentation as tabs

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,111 @@ function htmlToText(htmlString) {
+					// link in share-menu
+					$('a[id^=' + verb + '-' + ident.toString() + ']' )
+						.addClass('active')
+						.attr('href', 'javascript:doActivityItemAction(' + ident + ', "' + verb + '", true )').change();
+					$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child' ).removeClass('fa-retweet').addClass('fa-ban');
+				}
+				$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass).show();
+				if (verb.indexOf('announce') === 0 ) {
+					// ShareMenuButton
+					$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').addClass('fa-share');
+					if (data.verb == 'un' + verb) {
+						$('button[id^=shareMenuOptions-' + ident.toString() + ']').removeClass('active');
+					} else {
+						$('button[id^=shareMenuOptions-' + ident.toString() + ']').addClass('active');
+					}
+
+				} 
+			} else {
+				/* server-response was not ok. Database-problems or some changes in
+				 * data?
+				 * reset all buttons
+				 */
+				$('img[id^=waitfor-' + verb + '-' + ident.toString() + ']').remove();
+				$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').addClass('fa-share');
+				$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
+				$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
+				$.jGrowl(verb + ' not successful (server error)', {sticky: false, theme: 'info', life: 5000});
+			}
+		})
+	.error(
+			function(data){
+				/* Server could not be reaches successfully */
+				// remindert to remove the like-rotator from templates
+				//$('#like-rotator-' + ident.toString()).hide();
+				$('img[id^=waitfor-' + verb + '-' + ident.toString() + ']').remove();
+				$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').addClass('fa-share');
+				$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
+				$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
+				$.jGrowl(verb + ' not successful (network error)', {sticky: false, theme: 'info', life: 5000});
+		});
```
You can save an indentation level by inlining the success and error callbacks.
```suggestion
	.success(function(data){
		//$('#like-rotator-' + ident.toString()).hide();
		$('img[id^=waitfor-' + verb + '-' + ident.toString() + ']').remove();
		if (data.status == 'ok') {
			// response from server was ok
			if (data.verb == 'un' + verb) {
				// like/dislike buttons
				$('button[id^=' + verb + '-' + ident.toString() + ']' )
					.removeClass('active')
					.attr('onclick', 'doActivityItemAction(' + ident +', "' + verb + '",false )').change();
				// link in share-menu
				$('a[id^=' + verb + '-' + ident.toString() + ']' )
					.removeClass('active')
					.attr('href', 'javascript:doActivityItemAction(' + ident +', "' + verb + '",false )').change();
				$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child' ).addClass('fa-retweet').removeClass('fa-ban');
			} else {
				// like/dislike buttons
				$('button[id^=' + verb + '-' + ident.toString() + ']' )
					.addClass('active')
					.attr('onclick', 'doActivityItemAction(' + ident + ', "' + verb + '", true )').change();
				// link in share-menu
				$('a[id^=' + verb + '-' + ident.toString() + ']' )
					.addClass('active')
					.attr('href', 'javascript:doActivityItemAction(' + ident + ', "' + verb + '", true )').change();
				$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child' ).removeClass('fa-retweet').addClass('fa-ban');
			}
			$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass).show();
			if (verb.indexOf('announce') === 0 ) {
				// ShareMenuButton
				$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').addClass('fa-share');
				if (data.verb == 'un' + verb) {
					$('button[id^=shareMenuOptions-' + ident.toString() + ']').removeClass('active');
				} else {
					$('button[id^=shareMenuOptions-' + ident.toString() + ']').addClass('active');
				}
			} 
		} else {
			/* server-response was not ok. Database-problems or some changes in
			 * data?
			 * reset all buttons
			 */
			$('img[id^=waitfor-' + verb + '-' + ident.toString() + ']').remove();
			$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').addClass('fa-share');
			$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
			$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
			$.jGrowl(verb + ' not successful (server error)', {sticky: false, theme: 'info', life: 5000});
		}
	})
	.error(function(data){
		/* Server could not be reaches successfully */
		// remindert to remove the like-rotator from templates
		//$('#like-rotator-' + ident.toString()).hide();
		$('img[id^=waitfor-' + verb + '-' + ident.toString() + ']').remove();
		$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').addClass('fa-share');
		$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
		$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
		$.jGrowl(verb + ' not successful (network error)', {sticky: false, theme: 'info', life: 5000});
	});
```

MrPetovan commented on view/theme/frio/js/theme.js:
```diff
@@ -764,12 +764,111 @@ function htmlToText(htmlString) {
+				$('button[id^=shareMenuOptions-' + ident.toString() + '] i:first-child').addClass('fa-share');
+				$('button[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
+				$('a[id^=' + verb + '-' + ident.toString() + '] i:first-child').addClass(thumbsClass);
+				$.jGrowl(verb + ' not successful (network error)', {sticky: false, theme: 'info', life: 5000});
```
This isn't ideal because we can't translate this message from within the Javascript file. Usually we would generate the error message from the asynchronous call but since this is an error state we can't do that. Instead, you have to initialize Javascript strings with all the possible error messages by verb in a template that is fed localized strings from the PHP module.

For frio, this is done in `view/theme/frio/templates/js_strings.tpl` and the localized strings are declared in `src/App/Page.php:245`.

Commit:
reduce indentation

Commit:
add personal README
